### ISSUE TITLE
Can't serve/deploy built Angular app because of symlink in web folder

### ISSUE THREAD

**Issue by chalin** [dart-lang/build#890]
@jacob314 @matanlurey - When using the new build system over an Angular example app, it creates the following project structure (assume `--output=build`):

```nocode
build/
  packages/
    ...
  test/
    packages -> ../packages (symlink)
    ...
  web/
    packages -> ../packages (symlink)
    ...
```
I used to be able to run Seth's simple [dhttp](https://pub.dartlang.org/packages/dhttpd) server as follows: `dhttpd --path build/web`. But this doesn't work anymore because of the **symlink that points outside of the web folder**.

This symlink is also causing problems when I publish to **GitHub pages**.

Can we avoid the symlink inside the web folder?

cc @kwalrath @kevmoo

**Comment by matanlurey** [comment]
@jakemac53 @natebosch WAI, or something we should act on?

**Comment by natebosch** [comment]
I think we probably need to add a way to output only one top level directory, like `--output=web:build` will not output the `build/test`, just `build/packages/...` and `build/main.dart` etc.

If we are outputting multiple top level directories and we don't do the symlinks then I think dartium wouldn't work because it will try to load `web/packages/something/foo.dart` and fail.

**Comment by chalin** [comment]
But dartium is out of the picture, no?

**Comment by natebosch** [comment]
> But dartium is out of the picture, no?

Mostly - but it technically works with this today.

I'm not sure how source_maps would look.

**Comment by chalin** [comment]
Although it isn't a use case I have now (since `pub run build_runner test` works), but in the past I've run `pub build web test --web-compiler=something --mode=debug` and then run angular tests by hand by running two separate servers, one over each folder (web and test). With what you propose (to build each folder web or test folder separately) would I loose that ability?

**Comment by chalin** [comment]
(Oops, sorry @jacob314, I meant to choose @jakemac53 in my opening comment :)

**Comment by matanlurey** [comment]
Talked offline with @natebosch here is what we're suggesting:

1. Use `build -o build` for deployment (or emulating deployment).
2. Use `test` to _run_ tests.

I see you also want to _debug_ (i.e. use a browser to view `test`), so we could add:

```bash
$ --output=web:build
```

If that is right @chalin let's update this issue to reflect that request.

**Comment by chalin** [comment]
Yes, that sounds right, thanks. So if I understand your proposal I would be able to write:

- `--output=web:web-build`, and
- `--output=test:test-build`

and then run servers off of the `web-build` and `test-build` folders (over, say, ports 8080 and 8081, respectively).

**Comment by jakemac53** [comment]
You should be able to run servers within the existing `web` or `test` folders in the main output directory, the symlinks themselves should be reachable. Is there a flag for dhttpd that makes it not follow symlinks, or possibly just allows it to serve files outside the dir it is serving? The latter is not exactly safe because it exposes your whole machine, but the former only exposes files that are explicitly symlinked and should be fine.

**Comment by chalin** [comment]
cc @sethladd - dhttpd doesn't support following symlinks AFAIK. I've been falling back on Node-based lightweight servers. (That doesn't solve deployment issues, but the solution you propose will address that.)

**Comment by kevmoo** [comment]
@grouma @natebosch @jakemac53 – let's make sure we discuss how cleanup will happen. I'm nervous about having a bunch of solutions – some of which are hidden

CC @nshahan

**Comment by natebosch** [comment]
@kevmoo - I'm not sure that we talked about "cleanup" in relation to this issue?

We did talk about it relation to https://github.com/dart-lang/build/issues/1010

**Comment by kevmoo** [comment]
@natebosch – if I build out the web directory `-o web:build` – I don't want all of the Dart files, etc.

Do we consider these orthogonal?

**Comment by natebosch** [comment]
Yes these are orthogonal. You can remove .dart files with `-o build` or you could leave them in with `-o web:build`. They'll have separate ways to configure your intention and unrelated implementations.

**Comment by kevmoo** [comment]
So this will build one directory, without symlinks – but also without cleanup?

**Comment by natebosch** [comment]
> So this will build one directory, without symlinks – but also without cleanup?

Yes, removing source .dart files and other artifacts is (still) orthogonal.
